landoapi/models/landing_job.py

```diff
@@ -0,0 +1,209 @@
+    error = db.Column(db.Text(), default="")
+
+    # LDAP email of the user who requested transplant.
+    requester_email = db.Column(
+        db.String(254),
+        db.CheckConstraint("requester_email <> ''", name="requester_email_not_empty"),
+        nullable=False,
+    )
+
+    # Lando's name for the repository.
+    repository_name = db.Column(
+        db.Text(),
+        db.CheckConstraint("repository_name <> ''", name="repository_name_not_empty"),
+        nullable=False,
+    )
+
+    # URL of the repository revisions are to land to.
+    repository_url = db.Column(db.Text(), default="")
+
+    # Identifier for the most descendent commit created by this landing.
+    landed_commit_id = db.Column(db.Text(), default="")
+
+    # JSON list of bug ids tied to the revisions at creation time.
+    bug_ids = db.Column(JSONB, nullable=False)
+
+    # Number of attempts made to complete the job.
+    attempts = db.Column(db.Integer, nullable=False, default=0)
+
+    # Priority of the job. Higher values are processed first.
+    priority = db.Column(db.Integer, nullable=False, default=0)
+
+    def __repr__(self):
+        return "<LandingJob: %s>" % self.id
+
+    @property
+    def landing_path(self):
+        return [(int(r), self.revision_to_diff_id[r]) for r in self.revision_order]
+
+    @property
+    def head_revision(self):
+        """Human-readable representation of the branch head's Phabricator revision ID.
+        """
+        assert (
+            self.revision_order
+        ), "head_revision should never be called without setting self.revision_order!"
+        return "D" + self.revision_order[-1]
+
+    @classmethod
+    def revisions_query(cls, revisions):
+        revisions = [str(int(r)) for r in revisions]
+        return cls.query.filter(cls.revision_to_diff_id.has_any(array(revisions)))
+
+    @classmethod
+    def job_queue_query(cls, repositories=None):
+        """Return a query which selects the queued jobs."""
+        q = cls.query.filter(
+            cls.status.in_([LandingJobStatus.IN_PROGRESS, LandingJobStatus.SUBMITTED])
+        )
+
+        if repositories:
+            q = q.filter(cls.repository_name.in_(repositories))
+
+        # Any `LandingJobStatus.IN_PROGRESS` job is first and there should
+        # be a maximum of one (per repository). For
+        # `LandingJobStatus.SUBMITTED` jobs, higher priority items come first
+        # and then we order by creation time (older first).
+        q = q.order_by(cls.status.desc(), cls.priority.desc(), cls.created_at)
+
+        return q
+
+    @classmethod
+    def next_job_for_update_query(cls, repositories=None):
+        """Return a query which selects the next job and locks the row."""
+        q = cls.job_queue_query(repositories=repositories)
+
+        # Returned rows should be locked for updating, this ensures the next
+        # job can be claimed.
+        q = q.with_for_update()
+
+        return q
+
+    def landed(self, commit_id):
```

Comment:
nit: Should this be renamed to `land`? I.e. when using it, `landing_job.land(<commit_id>)` would make a little more sense than this. Same with `failed_transient` and `failed_permanent` below.